internal/pkg/addon/package.go

```diff
@@ -0,0 +1,243 @@
+		{
+			Property: []string{"RequestMappingTemplateS3Location"},
+		},
+		{
+			Property: []string{"ResponseMappingTemplateS3Location"},
+		},
+	},
+	"AWS::AppSync::FunctionConfiguration": {
+		{
+			Property: []string{"RequestMappingTemplateS3Location"},
+		},
+		{
+			Property: []string{"ResponseMappingTemplateS3Location"},
+		},
+	},
+	"AWS::ElasticBeanstalk::ApplicationVersion": {
+		{
+			Property:           []string{"SourceBundle"},
+			BucketNameProperty: "S3Bucket",
+			ObjectKeyProperty:  "S3Key",
+		},
+	},
+	"AWS::Glue::Job": {
+		{
+			Property: []string{"Command", "ScriptLocation"},
+		},
+	},
+	"AWS::StepFunctions::StateMachine": {
+		{
+			Property:           []string{"DefinitionS3Location"},
+			BucketNameProperty: "Bucket",
+			ObjectKeyProperty:  "Key",
+		},
+	},
+	"AWS::CodeCommit::Repository": {
+		{
+			Property:           []string{"Code", "S3"},
+			BucketNameProperty: "Bucket",
+			ObjectKeyProperty:  "Key",
+			ForceZip:           true,
+		},
+	},
+}
+
+func (a *Addons) packageTemplate(tmpl *cfnTemplate) (*cfnTemplate, error) {
+	resources := mappingNode(&tmpl.Resources)
+
+	for name, node := range resources {
+		resType := yamlMapGet(node, "Type").Value
+		confs, ok := resourcePackageConfig[resType]
+		if !ok {
+			continue
+		}
+
+		props := yamlMapGet(node, "Properties")
+		for _, conf := range confs {
+			if err := a.packageProperty(props, conf); err != nil {
+				return nil, fmt.Errorf("transform property %s property for %s: %w", name, conf.Property, err)
+			}
+		}
+	}
+
+	return tmpl, nil
+}
+
+// yamlMapGet parses node as a yaml map and searches key. If found,
+// it returns the value node of key. If node is not a yaml MappingNode
+// or key is not in the map, a zero value yaml.Node is returned, not a nil value,
+// to avoid panics and simplify accessing values from the returned node.
+//
+// If you need access many values from yaml map, consider mappingNode() instead, as
+// yamlMapGet will iterate through keys in the map each time vs a constant lookup.
+func yamlMapGet(node *yaml.Node, key string) *yaml.Node {
+	if node == nil || node.Kind != yaml.MappingNode {
+		return &yaml.Node{}
+	}
+
+	for i := 0; i < len(node.Content); i += 2 {
+		if node.Content[i].Value == key {
+			return node.Content[i+1]
+		}
+	}
+
+	return &yaml.Node{}
+}
+
+func (a *Addons) packageProperty(properties *yaml.Node, conf packagePropertyConfig) error {
+	node := properties
+	for _, key := range conf.Property {
+		node = yamlMapGet(node, key)
+	}
+
+	if node.IsZero() || node.Kind != yaml.ScalarNode {
+		// only transform if the node is a scalar node
+		return nil
+	}
+
+	if !isFilePath(node.Value) {
+		return nil
+	}
+
+	obj, err := a.uploadAddonAsset(node.Value, conf.ForceZip)
+	if err != nil {
+		return fmt.Errorf("upload asset: %w", err)
+	}
+
+	if len(conf.BucketNameProperty) == 0 && len(conf.ObjectKeyProperty) == 0 {
+		node.Value = s3.Location(obj.Bucket, obj.Key)
+		return nil
+	}
+
+	return node.Encode(map[string]string{
```

Comment:
nice! so this will replace the existing `node.Value`?

Reply:
correct! ☺️